experimental/list_formatter/src/list_formatter.rs

```diff
@@ -0,0 +1,223 @@
+            SimpleFormattedStringBuilder::<FieldType>::new,
+            |value| {
+                let mut builder = SimpleFormattedStringBuilder::<FieldType>::new();
+                builder.append(value, FieldType::Element);
+                builder
+            },
+            |value, (before, between, after), mut builder| {
+                builder.append(after, FieldType::Literal);
+                builder.prepend(between, FieldType::Literal);
+                builder.prepend(value, FieldType::Element);
+                builder.prepend(before, FieldType::Literal);
+                builder
+            }
+        )
+    }
+}
+
+type PatternParts<'a> = (&'a str,&'a str,&'a str);
+enum Pattern<'a> {
+    Simple(PatternParts<'a>),
+    Conditional { cond: Regex, then: Box<Pattern<'a>>, else_: Box<Pattern<'a>> },
+}
+
+impl <'a> Pattern<'a> {
+    fn parse(raw: &'a str) -> Pattern<'a> {
+        if raw.starts_with("{cond}") {
+            let then_index = raw.find("{then}").expect("missing {then}");
+            let else_index = raw.find("{else}").expect("missing {else}");
+            return Pattern::Conditional{
+                cond: Regex::new(&raw[6..then_index]).expect("invalid regex"), 
+                then: Box::new(Pattern::parse(&raw[then_index+6..else_index])), 
+                else_: Box::new(Pattern::parse(&raw[else_index+6..]))
+            };
+        }
+        let index_0 = raw.find("{0}").expect("missing {0}");
+        let index_1 = raw.find("{1}").expect("missing {1}");
+        Pattern::Simple((&raw[0..index_0], &raw[index_0+3..index_1], &raw[index_1+3..]))
+    }
+
+    fn get_parts(&self, following_value: &str) -> &PatternParts<'a> {
+        match self {
+            Pattern::Simple(parts) => parts,
+            Pattern::Conditional{cond, then, else_} => 
+                if cond.is_match(following_value) {then.get_parts(following_value)} else {else_.get_parts(following_value)},
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const VALUES: &[&str] = &["one", "two", "three", "four", "five"];
+
+    fn test_formatter() -> ListFormatter<'static> {
+        ListFormatter {
+            pair: Pattern::parse("{0}; {1}"),
+            first: Pattern::parse("{0}: {1}"),
+            middle: Pattern::parse("{0}, {1}"),
+            last: Pattern::parse("{0}. {1}!"),
+        }
+    }
+
+    #[test]
+    fn test_format() {
+        assert_eq!(test_formatter().format(&VALUES[0..0]), "");
+        assert_eq!(test_formatter().format(&VALUES[0..1]), "one");
+        assert_eq!(test_formatter().format(&VALUES[0..2]), "one; two");
+        assert_eq!(test_formatter().format(&VALUES[0..3]), "one: two. three!");
+        assert_eq!(
+            test_formatter().format(&VALUES[0..4]),
+            "one: two, three. four!"
+        );
+        assert_eq!(
+            test_formatter().format(VALUES),
+            "one: two, three, four. five!"
+        );
+    }
+
+    #[test]
+    fn test_format_to_parts() {
```

Comment:
Suggestion: Make a data-driven test utility for FormattedStringBuilder that can be shared across components everywhere it is needed.

Reply:
Not sure I follow

Reply:
Make a note to work on this in a future commit.